libs/metrics/social_distancing.py

```diff
@@ -0,0 +1,187 @@
+import csv
+import ast
+import numpy as np
+import os
+
+from collections import deque
+from datetime import datetime, date, timedelta
+from typing import Dict, List, Iterator, Tuple
+
+from libs.utils.loggers import get_source_log_directory
+
+from .base import BaseMetric
+
+
+class SocialDistancingMetric(BaseMetric):
+
+    reports_folder = "social-distancing"
+    csv_headers = ["DetectedObjects", "NoInfringement", "LowInfringement", "HighInfringement",
+                   "CriticalInfringement"]
+
+    @classmethod
+    def procces_csv_row(cls, csv_row: Dict, objects_logs: Dict):
+        row_time = datetime.strptime(csv_row["Timestamp"], "%Y-%m-%d %H:%M:%S")
+        detections = ast.literal_eval(csv_row["Detections"])
+        row_hour = row_time.hour
+        if not objects_logs.get(row_hour):
+            objects_logs[row_hour] = {}
+        for index, d in enumerate(detections):
+            if not objects_logs[row_hour].get(d["tracking_id"]):
+                objects_logs[row_hour][d["tracking_id"]] = {"distance_violations": []}
+            # Append social distancing violations
+            objects_logs[row_hour][d["tracking_id"]]["distance_violations"].append(
+                {
+                    "time": row_time,
+                    "infrigement": index in ast.literal_eval(csv_row["ViolationsIndexes"])
+                }
+            )
+
+    @classmethod
+    def generate_hourly_metric_data(cls, objects_logs):
+        summary = np.zeros((len(objects_logs), 5), dtype=np.long)
+        for index, hour in enumerate(sorted(objects_logs)):
+            hour_objects_detections = objects_logs[hour]
+            for detection_object in hour_objects_detections.values():
+                summary[index] += cls.process_distance_violation_for_object(
+                    detection_object["distance_violations"])
+        return summary
+
+    @classmethod
+    def process_distance_violation_for_object(cls, distance_violations: List[dict]) -> Tuple[int, int]:
+        """
+        Receives a list with the "social distancing detections" (for a single person) and returns a
+        tuple with the summary of detections and violations (grouped by severity). Consecutive detections in
+        the same state are grouped and returned as a single one. Detections lower than the constant
+        PROCESSING_COUNT_THRESHOLD are ignored.
+
+        The infrigement categories are :
+            - Low: Between 10 seconds 30 seconds
```

Comment:
I know these numbers can be changed but where did we get them from? If the idea is to avoid COVID then we should base it on research and less than 5 or maybe even 15 minutes should be low. Critical I would say could be something like 2 hours. But we should validate those numbers

Reply:
I added a comment mentioning that the values only have demo purposes a needs to be checked with the OMS recommendations. If we use the real values, most of the video demo that we have will only generate low infractions.